File: apps/server/api/src/terminal/pty.manager.ts
/**
 * Terminal PTY Manager.
 *
 * Owns the lifecycle of every terminal session on the server, backed by
 * Bun.Terminal (native PTY built into the Bun runtime). A PTY is identified
 * by a random `terminalId`; the WebSocket route creates instances here and
 * streams their output/exit events back to the client.
 *
 * All instances are tracked so a server shutdown (or daemon restart) can kill
 * lingering shells deterministically instead of leaking processes.
 */
import { randomUUID } from "node:crypto";
import { existsSync, statSync } from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import type {
  TerminalId,
  TerminalOutputEvent,
  TerminalSpawnParams,
  TerminalSpawnedEvent,
} from "@console/types";

/** Callback the route registers to receive pty events for a session. */
export interface PtyCallbacks {
  onData: (event: TerminalOutputEvent) => void;
  onExit: (code: number | null) => void;
  onError: (message: string) => void;
}

/** Minimal shape of a Bun-spawned subprocess attached to a terminal. */
interface PtyProcess {
  pid: number;
  exited: Promise<number>;
  kill(code?: number): void;
}

/** Cap for output buffered while paused, so a flooding program can't balloon memory. */
const PAUSED_BUFFER_LIMIT_BYTES = 8 * 1024 * 1024;

/** Coalescing window: PTY data events inside one window ship as a single frame. */
const OUTPUT_FLUSH_MS = 8;
/** Flush early once the queued window reaches this size. */
const OUTPUT_FLUSH_BYTES = 4 * 1024;
/** Max bytes per WebSocket frame when flushing a large backlog. */
const OUTPUT_FRAME_BYTES = 64 * 1024;

interface PtySession {
  id: TerminalId;
  terminal: Bun.Terminal;
  /** null until the shell subprocess actually starts (see startShell). */
  proc: PtyProcess | null;
  shell: string;
  cwd: string;
  cols: number;
  rows: number;
  callbacks?: PtyCallbacks;
  pending: string[];
  killed: boolean;
  paused: boolean;
  pausedBuffer: string[];
  pausedBufferBytes: number;
  /** Output chunks waiting for the coalescing flush (see OUTPUT_FLUSH_MS). */
  outputQueue: string[];
  outputQueueBytes: number;
  flushTimer?: ReturnType<typeof setTimeout>;
  decoder: TextDecoder;
  /** Whether the shell subprocess has been started. */
  shellStarted: boolean;
  /** Input received before the shell started, flushed on start. */
  pendingInput: string[];
  resolveStart?: (event: TerminalSpawnedEvent) => void;
  rejectStart?: (cause: unknown) => void;
}

// --- Hardening: PTY spawn allowlist + rate limits ---

const ALLOWED_SHELLS = new Set<string>([
  "/bin/bash",
  "/bin/zsh",
  "/bin/sh",
  "/usr/bin/bash",
  "/usr/bin/zsh",
  "/usr/local/bin/bash",
  "/usr/local/bin/zsh",
  "/bin/fish",
  "/usr/bin/fish",
  "powershell.exe",
  "pwsh.exe",
  "cmd.exe",
]);

const MAX_CONCURRENT_TERMINALS = 20;
const MAX_SPAWNS_PER_MINUTE = 20;

function isAllowedShell(shell: string): boolean {
  if (ALLOWED_SHELLS.has(shell)) return true;
  const hostShell = process.env.SHELL;
  if (hostShell && shell === hostShell) return true;
  const base = path.basename(shell);
  for (const allowed of ALLOWED_SHELLS) {
    if (path.basename(allowed) === base) return true;
  }
  return false;
}

export class TerminalPtyManager {
  private sessions = new Map<TerminalId, PtySession>();
  private spawnTimestamps: number[] = [];

  /**
   * Spawn a new shell PTY in the given working directory.
   * Throws if `cwd` does not exist so clients get a clean error.
   *
   * The shell starts immediately at the requested size — clients already send
   * their true grid in the spawn params, so waiting for a first resize only
   * adds latency to every open.
   */
  spawn(params: TerminalSpawnParams): {
    id: TerminalId;
    ready: Promise<TerminalSpawnedEvent>;
  } {
    // Rate-limit + concurrency guard before allocating any FD.
    const now = Date.now();
    this.spawnTimestamps = this.spawnTimestamps.filter((t) => now - t < 60_000);
    if (this.spawnTimestamps.length >= MAX_SPAWNS_PER_MINUTE) {
      throw new Error("Too many terminal spawns — rate limited. Try again later.");
    }
    if (this.sessions.size >= MAX_CONCURRENT_TERMINALS) {
      throw new Error("Too many concurrent terminals.");
    }

    const cwd = path.resolve(params.cwd);
    if (!existsSync(cwd)) {
      throw new Error(`Cannot spawn terminal: working directory does not exist: ${cwd}`);
    }
    try {
      const st = statSync(cwd);
      if (!st.isDirectory()) throw new Error(`Not a directory: ${cwd}`);
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      if (msg.startsWith("Not a directory")) throw e;
      throw new Error(`Cannot spawn terminal: working directory is not a directory: ${cwd}`);
    }

    const rawShell = params.shell || process.env.SHELL || this.defaultShell();
    const shellForCheck = rawShell.includes("/") ? path.resolve(rawShell) : rawShell;
    const candidate = rawShell.includes("/") ? shellForCheck : rawShell;
    if (!isAllowedShell(candidate) && !isAllowedShell(path.basename(candidate))) {
      throw new Error(`Shell not allowed: ${rawShell}`);
    }
    const shell = candidate;
    const cols = params.cols ?? 80;
    const rows = params.rows ?? 24;

    const id: TerminalId = randomUUID();
    const decoder = new TextDecoder("utf-8", { fatal: false });
    const session: PtySession = {
      id,
      // Assigned immediately after construction below.
      terminal: undefined as unknown as Bun.Terminal,
      // Assigned in startShell(); null until then.
      proc: null,
      shell,
      cwd,
      cols,
      rows,
      pending: [],
      killed: false,
      paused: false,
      pausedBuffer: [],
      pausedBufferBytes: 0,
      outputQueue: [],
      outputQueueBytes: 0,
      decoder,
      shellStarted: false,
      pendingInput: [],
    };
    this.sessions.set(id, session);
    this.spawnTimestamps.push(Date.now());

    session.terminal = new Bun.Terminal({
      name: "xterm-256color",
      cols,
      rows,
      // Buffer output that arrives before the WebSocket route has attached
      // callbacks so the initial shell prompt is never dropped.
      data: (_terminal, data) => {
        const text = session.decoder.decode(data, { stream: true });
        if (text.length > 0) {
          this.handleOutput(session, text);
        }
      },
    });

    const ready = new Promise<TerminalSpawnedEvent>((resolve, reject) => {
      session.resolveStart = resolve;
      session.rejectStart = reject;
    });
    // Start the shell right away at the requested grid size.
    this.startShell(session);

    return { id, ready };
  }

  /** Start the shell subprocess and resolve the spawn promise. */
  private startShell(session: PtySession): void {
    if (session.shellStarted || session.killed) return;
    session.shellStarted = true;

    const proc = Bun.spawn([session.shell], {
      terminal: session.terminal,
      cwd: session.cwd,
      env: {
        ...process.env,
        TERM: "xterm-256color",
        CONSOLE_TERMINAL: "true",
      } as Record<string, string>,
    });
    session.proc = {
      pid: proc.pid,
      exited: proc.exited,
      kill: (code?: number) => proc.kill(code),
    };

    proc.exited.then((code) => {
      if (session.killed) return;
      session.killed = true;
      this.sessions.delete(session.id);
      session.callbacks?.onExit(code);
    });

    // Flush any keystrokes that arrived before the shell existed.
    if (session.pendingInput.length > 0) {
      const queued = session.pendingInput.join("");
      session.pendingInput.length = 0;
      try {
        session.terminal.write(queued);
      } catch {
        // Terminal closed mid-flush — nothing more to do.
      }
    }

    session.resolveStart?.({
      type: "spawned",
      id: session.id,
      pid: proc.pid,
      shell: session.shell,
      cwd: session.cwd,
      cols: session.cols,
      rows: session.rows,
    });
  }

  /** Route PTY output to callbacks, coalescing bursts into fewer frames. */
  private handleOutput(session: PtySession, data: string): void {
    if (session.killed) return;
    if (session.paused) {
      // Client send buffer saturated: hold output until resume().
      session.pausedBuffer.push(data);
      session.pausedBufferBytes += data.length;
      if (session.pausedBufferBytes > PAUSED_BUFFER_LIMIT_BYTES) {
        // Drop oldest to stay bounded; the client is already behind anyway.
        const dropped = session.pausedBuffer.shift()!;
        session.pausedBufferBytes -= dropped.length;
      }
      return;
    }
    if (!session.callbacks) {
      session.pending.push(data);
      // Cap buffered early output to avoid unbounded growth if attach never happens
      if (session.pending.length > 100) {
        session.pending.shift();
      }
      return;
    }
    session.outputQueue.push(data);
    session.outputQueueBytes += data.length;
    if (session.outputQueueBytes >= OUTPUT_FLUSH_BYTES) {
      this.flushOutput(session);
    } else if (!session.flushTimer) {
      session.flushTimer = setTimeout(() => this.flushOutput(session), OUTPUT_FLUSH_MS);
    }
  }

  /** Ship the coalesced output window as a single frame. */
  private flushOutput(session: PtySession): void {
    if (session.flushTimer) {
      clearTimeout(session.flushTimer);
      session.flushTimer = undefined;
    }
    if (session.outputQueue.length === 0 || session.killed) return;
    const queued = session.outputQueue;
    session.outputQueue = [];
    session.outputQueueBytes = 0;
    if (session.paused || !session.callbacks) {
      // Paused (or detached) mid-window: hold for resume/attach instead.
      for (const chunk of queued) {
        session.pausedBuffer.push(chunk);
        session.pausedBufferBytes += chunk.length;
      }
      return;
    }
    session.callbacks.onData({ type: "output", data: queued.join("") });
  }

  /** Attach a WebSocket-backed callback set to an existing session. */
  attach(id: TerminalId, callbacks: PtyCallbacks): void {
    const session = this.sessions.get(id);
    if (!session) throw new Error(`No terminal session found for id: ${id}`);
    session.callbacks = callbacks;
    // Flush any output that arrived between spawn and attach (e.g. shell prompt)
    if (session.pending.length > 0) {
      const queued = session.pending.join("");
      session.pending.length = 0;
      callbacks.onData({ type: "output", data: queued });
    }
  }

  /** Write raw bytes into the PTY (keystrokes, pasted text). */
  write(id: TerminalId, data: string): boolean {
    if (data.length > 256 * 1024) return false;
    const session = this.sessions.get(id);
    if (!session || session.killed) return false;
    // Shell not started yet (waiting for first resize): hold keystrokes.
    if (!session.shellStarted) {
      session.pendingInput.push(data);
      return true;
    }
    try {
      session.terminal.write(data);
      return true;
    } catch {
      return false;
    }
  }

  /** Resize the PTY's viewport. */
  resize(id: TerminalId, cols: number, rows: number): boolean {
    if (!Number.isFinite(cols) || !Number.isFinite(rows) || cols < 1 || cols > 500 || rows < 1 || rows > 200) return false;
    const session = this.sessions.get(id);
    if (!session || session.killed) return false;
    try {
      session.terminal.resize(cols, rows);
      session.cols = cols;
      session.rows = rows;
      return true;
    } catch {
      return false;
    }
  }

  /** Kill a PTY session and remove it from the registry. */
  kill(id: TerminalId): void {
    const session = this.sessions.get(id);
    if (!session || session.killed) return;
    session.killed = true;
    this.sessions.delete(id);
    if (session.flushTimer) {
      clearTimeout(session.flushTimer);
      session.flushTimer = undefined;
    }
    if (!session.shellStarted) {
      session.rejectStart?.(new Error("Terminal killed before the shell started."));
    }
    try {
      // Interactive shells ignore SIGTERM/SIGHUP; SIGKILL is deterministic.
      session.proc?.kill(9);
    } catch {
      // Already dead — fine.
    }
    try {
      session.terminal.close();
    } catch {
      // Already closed — fine.
    }
    session.callbacks?.onExit(null);
  }

  /**
   * Pause a PTY's output. Used when the client socket's send buffer is
   * saturated so a flooding program (`yes`, huge file cats) can't balloon
   * memory; output is held here until resume(). Unlike node-pty's kernel-level
   * flow control this buffers in-process, capped at PAUSED_BUFFER_LIMIT_BYTES.
   */
  pause(id: TerminalId): void {
    const session = this.sessions.get(id);
    if (session && !session.killed) {
      session.paused = true;
    }
  }

  /** Resume a paused PTY's output (the send buffer drained). */
  resume(id: TerminalId): void {
    const session = this.sessions.get(id);
    if (session && !session.killed && session.paused) {
      session.paused = false;
      if (session.pausedBuffer.length === 0) return;
      const joined = session.pausedBuffer.join("");
      session.pausedBuffer = [];
      session.pausedBufferBytes = 0;
      // Split huge backlogs so no single frame balloons memory downstream.
      for (let i = 0; i < joined.length; i += OUTPUT_FRAME_BYTES) {
        this.handleOutput(session, joined.slice(i, i + OUTPUT_FRAME_BYTES));
      }
    }
  }

  /** Kill every tracked session (e.g. on server shutdown). */
  killAll(): void {
    for (const id of [...this.sessions.keys()]) {
      this.kill(id);
    }
  }

  get size(): number {
    return this.sessions.size;
  }

  private defaultShell(): string {
    if (os.platform() === "win32") return "powershell.exe";
    return "/bin/bash";
  }
}

/** Singleton used across the terminal route. */
export const terminalPtyManager = new TerminalPtyManager();

File: apps/server-go/internal/services/pty_manager.go
// PTY manager for the terminal WebSocket. Port of
// apps/server/api/src/terminal/pty.manager.ts (JSON + binary protocol).
//
// Performance hardening mirrors the Bun version:
//   - output coalescing (8ms window / 4KB early-flush / 64KB frame cap)
//   - bounded queues with oldest-drop (8MB cap) so flooding programs can't OOM
//   - pause/resume for send-buffer backpressure
//   - cached base env, shell allowlist, concurrency cap, input size cap
package services

import (
	"fmt"
	"os"
	"os/exec"
	"path"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/types"
	"github.com/creack/pty"
)

const maxSpawnsPerMinute = 30

// Coalescing window: PTY data inside one window ships as fewer frames.
const (
	outputFlushMs    = 8 * time.Millisecond
	outputFlushBytes = 4 * 1024
	outputFrameBytes = 64 * 1024
)

// Cap for buffered output so a flooding program can't balloon memory.
const pausedBufferLimitBytes = 8 * 1024 * 1024

// Max input bytes per Write (mirrors Bun's 256KB guard).
const maxTerminalInputBytes = 256 * 1024

const maxConcurrentTerminals = 20

// Shell allowlist mirroring Bun's ALLOWED_SHELLS.
var allowedShells = map[string]struct{}{
	"/bin/bash":          {},
	"/bin/zsh":           {},
	"/bin/sh":            {},
	"/usr/bin/bash":      {},
	"/usr/bin/zsh":       {},
	"/usr/local/bin/bash": {},
	"/usr/local/bin/zsh": {},
	"/bin/fish":          {},
	"/usr/bin/fish":      {},
	"powershell.exe":     {},
	"pwsh.exe":           {},
	"cmd.exe":            {},
}

func isAllowedShell(shell string) bool {
	if _, ok := allowedShells[shell]; ok {
		return true
	}
	if hostShell := os.Getenv("SHELL"); hostShell != "" && shell == hostShell {
		return true
	}
	base := path.Base(shell)
	for allowed := range allowedShells {
		if path.Base(allowed) == base {
			return true
		}
	}
	return false
}

// Cached base env: spawns are frequent enough that re-copying os.Environ per
// spawn is pure GC pressure — the server env is static after boot.
var (
	cachedBaseEnvOnce sync.Once
	cachedBaseEnv     []string
)

func baseEnv() []string {
	cachedBaseEnvOnce.Do(func() {
		cachedBaseEnv = os.Environ()
	})
	return cachedBaseEnv
}

type PtySession struct {
	ID  string
	cmd *exec.Cmd
	ptmx *os.File

	manager *PtyManager

	// Callbacks registered by the WS route. Guarded by outputMu.
	onData func([]byte)
	onExit func(int)

	// killed is set once; checked atomically so Write/Resize never hold a
	// mutex across blocking ptmx syscalls.
	killed atomic.Bool
	// writeMu serializes ptmx Write vs Setsize.
	writeMu sync.Mutex

	// Output coalescing + backpressure state, all guarded by outputMu.
	outputMu      sync.Mutex
	outputQueue   [][]byte
	outputBytes   int
	flushTimer    *time.Timer
	paused        bool
	pausedBuffer  [][]byte
	pausedBytes   int
	pending       [][]byte
	pendingCount  int
}

// SetCallbacks attaches the WS pump and flushes early output that arrived
// between spawn and attach (e.g. the shell prompt).
func (s *PtySession) SetCallbacks(onData func([]byte), onExit func(int)) {
	var toFlush [][]byte
	s.outputMu.Lock()
	s.onData = onData
	s.onExit = onExit
	if len(s.pending) > 0 {
		toFlush = s.pending
		s.pending = nil
		s.pendingCount = 0
	}
	s.outputMu.Unlock()

	for _, chunk := range toFlush {
		onData(chunk)
	}
}

func (s *PtySession) Write(data []byte) error {
	if s.killed.Load() {
		return fmt.Errorf("session killed")
	}
	if len(data) > maxTerminalInputBytes {
		return fmt.Errorf("input too large")
	}
	// No mutex held across the blocking write; killed is atomic so Kill
	// never blocks behind a slow ptmx.
	s.writeMu.Lock()
	defer s.writeMu.Unlock()
	if s.killed.Load() {
		return fmt.Errorf("session killed")
	}
	_, err := s.ptmx.Write(data)
	return err
}

// Pid returns the spawned shell's process id (0 when unknown). The desktop
// requires pid in the `spawned` frame to flip a terminal to Running.
// cmd is assigned once during Spawn before the session is shared.
func (s *PtySession) Pid() int {
	if s.cmd == nil || s.cmd.Process == nil {
		return 0
	}
	return s.cmd.Process.Pid
}

// Resize updates the PTY window size.
func (s *PtySession) Resize(cols, rows int) error {
	if cols < 1 || cols > 500 || rows < 1 || rows > 200 {
		return fmt.Errorf("invalid size")
	}
	if s.killed.Load() {
		return fmt.Errorf("session killed")
	}
	s.writeMu.Lock()
	defer s.writeMu.Unlock()
	if s.killed.Load() {
		return fmt.Errorf("session killed")
	}
	return pty.Setsize(s.ptmx, &pty.Winsize{Cols: uint16(cols), Rows: uint16(rows)})
}

func (s *PtySession) Kill() {
	if s.killed.Swap(true) {
		return
	}
	s.outputMu.Lock()
	if s.flushTimer != nil {
		s.flushTimer.Stop()
		s.flushTimer = nil
	}
	// Drop buffered output so a killed session releases memory promptly.
	s.outputQueue = nil
	s.outputBytes = 0
	s.pausedBuffer = nil
	s.pausedBytes = 0
	s.pending = nil
	s.outputMu.Unlock()

	if s.cmd.Process != nil {
		_ = s.cmd.Process.Kill()
	}
	_ = s.ptmx.Close()
}

// Pause holds output in a bounded buffer until Resume.
func (s *PtySession) Pause() {
	s.outputMu.Lock()
	defer s.outputMu.Unlock()
	if !s.killed.Load() {
		s.paused = true
	}
}

// Resume drains the held buffer back through the coalescing queue.
func (s *PtySession) Resume() {
	s.outputMu.Lock()
	if !s.paused || s.killed.Load() {
		s.outputMu.Unlock()
		return
	}
	s.paused = false
	held := s.pausedBuffer
	s.pausedBuffer = nil
	s.pausedBytes = 0
	s.outputMu.Unlock()

	for _, chunk := range held {
		s.enqueueOutput(chunk)
	}
}

// enqueueOutput routes one PTY read into the coalescing queue or the paused
// buffer. The caller must pass an owned copy (the read buffer is reused).
func (s *PtySession) enqueueOutput(chunk []byte) {
	s.outputMu.Lock()
	if s.killed.Load() {
		s.outputMu.Unlock()
		return
	}
	// No callbacks yet (spawn → attach gap): hold early output, capped.
	if s.onData == nil && !s.paused {
		cp := make([]byte, len(chunk))
		copy(cp, chunk)
		s.pending = append(s.pending, cp)
		s.pendingCount++
		if len(s.pending) > 100 {
			dropped := s.pending[0]
			s.pending = s.pending[1:]
			_ = dropped
			s.pendingCount--
		}
		s.outputMu.Unlock()
		return
	}
	if s.paused {
		cp := make([]byte, len(chunk))
		copy(cp, chunk)
		s.pausedBuffer = append(s.pausedBuffer, cp)
		s.pausedBytes += len(cp)
		for s.pausedBytes > pausedBufferLimitBytes && len(s.pausedBuffer) > 0 {
			dropped := s.pausedBuffer[0]
			s.pausedBuffer = s.pausedBuffer[1:]
			s.pausedBytes -= len(dropped)
		}
		s.outputMu.Unlock()
		return
	}
	cp := make([]byte, len(chunk))
	copy(cp, chunk)
	s.outputQueue = append(s.outputQueue, cp)
	s.outputBytes += len(cp)
	for s.outputBytes > pausedBufferLimitBytes && len(s.outputQueue) > 0 {
		dropped := s.outputQueue[0]
		s.outputQueue = s.outputQueue[1:]
		s.outputBytes -= len(dropped)
	}
	shouldFlush := s.outputBytes >= outputFlushBytes
	if shouldFlush {
		if s.flushTimer != nil {
			s.flushTimer.Stop()
			s.flushTimer = nil
		}
	} else if s.flushTimer == nil {
		s.flushTimer = time.AfterFunc(outputFlushMs, func() { s.flushOutput() })
	}
	s.outputMu.Unlock()

	if shouldFlush {
		s.flushOutput()
	}
}

// flushOutput ships the coalesced window as 64KB-capped frames.
func (s *PtySession) flushOutput() {
	s.outputMu.Lock()
	if s.flushTimer != nil {
		s.flushTimer.Stop()
		s.flushTimer = nil
	}
	if len(s.outputQueue) == 0 || s.killed.Load() {
		s.outputMu.Unlock()
		return
	}
	if s.paused || s.onData == nil {
		// Paused (or detached) mid-window: hold for resume/attach.
		for _, chunk := range s.outputQueue {
			s.pausedBuffer = append(s.pausedBuffer, chunk)
			s.pausedBytes += len(chunk)
		}
		for s.pausedBytes > pausedBufferLimitBytes && len(s.pausedBuffer) > 0 {
			dropped := s.pausedBuffer[0]
			s.pausedBuffer = s.pausedBuffer[1:]
			s.pausedBytes -= len(dropped)
		}
		s.outputQueue = nil
		s.outputBytes = 0
		s.outputMu.Unlock()
		return
	}
	queued := s.outputQueue
	s.outputQueue = nil
	s.outputBytes = 0
	onData := s.onData
	s.outputMu.Unlock()

	var joined []byte
	if len(queued) == 1 {
		joined = queued[0]
	} else {
		total := 0
		for _, c := range queued {
			total += len(c)
		}
		joined = make([]byte, 0, total)
		for _, c := range queued {
			joined = append(joined, c...)
		}
	}
	for i := 0; i < len(joined); i += outputFrameBytes {
		end := i + outputFrameBytes
		if end > len(joined) {
			end = len(joined)
		}
		onData(joined[i:end])
	}
}

type PtyManager struct {
	mu         sync.Mutex
	sessions   map[string]*PtySession
	spawnTimes []int64
}

func NewPtyManager() *PtyManager {
	return &PtyManager{sessions: make(map[string]*PtySession)}
}

// Spawn validates params and starts the shell under a PTY.
func (m *PtyManager) Spawn(params types.TerminalSpawnParams) (*PtySession, error) {
	if strings.Contains(params.Cwd, "\x00") {
		return nil, fmt.Errorf("Invalid cwd")
	}
	if info, err := os.Stat(params.Cwd); err != nil {
		return nil, fmt.Errorf("Cannot spawn terminal: working directory does not exist: %s", params.Cwd)
	} else if !info.IsDir() {
		return nil, fmt.Errorf("Cannot spawn terminal: working directory is not a directory: %s", params.Cwd)
	}

	m.mu.Lock()
	now := nowMillis()
	recent := m.spawnTimes[:0]
	for _, t := range m.spawnTimes {
		if now-t < 60_000 {
			recent = append(recent, t)
		}
	}
	m.spawnTimes = recent
	if len(m.spawnTimes) >= maxSpawnsPerMinute {
		m.mu.Unlock()
		return nil, fmt.Errorf("Too many terminal spawns — rate limited. Try again later.")
	}
	if len(m.sessions) >= maxConcurrentTerminals {
		m.mu.Unlock()
		return nil, fmt.Errorf("Too many concurrent terminals.")
	}
	m.spawnTimes = append(m.spawnTimes, now)
	m.mu.Unlock()

	shell := params.Shell
	if shell == "" {
		shell = os.Getenv("SHELL")
	}
	if shell == "" {
		shell = "/bin/bash"
	}
	if strings.Contains(shell, "\x00") {
		return nil, fmt.Errorf("Invalid shell")
	}
	if !isAllowedShell(shell) && !isAllowedShell(path.Base(shell)) {
		return nil, fmt.Errorf("Shell not allowed: %s", shell)
	}

	cols := params.Cols
	if cols <= 0 {
		cols = 80
	}
	rows := params.Rows
	if rows <= 0 {
		rows = 24
	}

	id := randomID()
	cmd := exec.Command(shell)
	cmd.Dir = params.Cwd
	base := baseEnv()
	env := make([]string, 0, len(base)+3)
	env = append(env, base...)
	env = append(env, "TERM=xterm-256color", "CONSOLE_TERMINAL=true", "CONSOLE_TERMINAL_ID="+id)
	cmd.Env = env

	ptmx, err := pty.StartWithSize(cmd, &pty.Winsize{Cols: uint16(cols), Rows: uint16(rows)})
	if err != nil {
		return nil, err
	}
	session := &PtySession{ID: id, cmd: cmd, ptmx: ptmx, manager: m}

	// Pump PTY output through the coalescing queue.
	go func() {
		buf := make([]byte, 32*1024)
		for {
			n, err := ptmx.Read(buf)
			if n > 0 {
				session.enqueueOutput(buf[:n])
			}
			if err != nil {
				break
			}
		}
		// Flush any coalesced output before reporting exit.
		session.flushOutput()
		code := 0
		if werr := cmd.Wait(); werr != nil {
			if exitErr, ok := werr.(*exec.ExitError); ok && exitErr.ExitCode() > 0 {
				code = exitErr.ExitCode()
			}
		}
		// Natural exit: remove from the registry so entries don't leak
		// under churn (explicit Kill also deletes; this covers the
		// process-exits-first path).
		if session.manager != nil {
			session.manager.remove(session.ID)
		}
		session.Kill()
		session.outputMu.Lock()
		onExit := session.onExit
		session.outputMu.Unlock()
		if onExit != nil {
			onExit(code)
		}
	}()

	m.mu.Lock()
	m.sessions[session.ID] = session
	m.mu.Unlock()
	return session, nil
}

func (m *PtyManager) remove(id string) {
	m.mu.Lock()
	delete(m.sessions, id)
	m.mu.Unlock()
}

func (m *PtyManager) Get(id string) *PtySession {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.sessions[id]
}

func (m *PtyManager) Kill(id string) {
	m.mu.Lock()
	session := m.sessions[id]
	delete(m.sessions, id)
	m.mu.Unlock()
	if session != nil {
		session.Kill()
	}
}

func (m *PtyManager) Pause(id string) {
	m.mu.Lock()
	session := m.sessions[id]
	m.mu.Unlock()
	if session != nil {
		session.Pause()
	}
}

func (m *PtyManager) Resume(id string) {
	m.mu.Lock()
	session := m.sessions[id]
	m.mu.Unlock()
	if session != nil {
		session.Resume()
	}
}

func (m *PtyManager) Write(id string, data []byte) bool {
	m.mu.Lock()
	session := m.sessions[id]
	m.mu.Unlock()
	if session == nil {
		return false
	}
	return session.Write(data) == nil
}

func (m *PtyManager) KillAll() {
	m.mu.Lock()
	sessions := make([]*PtySession, 0, len(m.sessions))
	for _, s := range m.sessions {
		sessions = append(sessions, s)
	}
	m.sessions = make(map[string]*PtySession)
	m.mu.Unlock()
	for _, s := range sessions {
		s.Kill()
	}
}
